src/tree/index.tsx

```diff
@@ -0,0 +1,415 @@
+}
+
+interface TreeCache {
+	activeNode?: string;
+	selectedNode?: string;
+	controlledExpandedNodes: string[];
+	controlledCheckedNodes: string[];
+	checked: string[];
+	expanded: string[];
+}
+
+const resource = createResourceMiddleware<TreeNodeOption>();
+const icache = createICacheMiddleware<TreeCache>();
+const factory = create({ theme, icache, diffProperty, focus, resource })
+	.properties<TreeProperties>()
+	.children<TreeChildren | undefined>();
+
+export default factory(function({
+	id,
+	middleware: { theme, icache, diffProperty, focus, resource },
+	properties,
+	children
+}) {
+	diffProperty(
+		'selectedNode',
+		properties,
+		({ selectedNode: current }, { selectedNode: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('selectedNode', next);
+			}
+		}
+	);
+	diffProperty('expanded', properties, ({ expanded: current }, { expanded: next }) => {
+		if ((current || next) && current !== next) {
+			icache.set('controlledExpandedNodes', next || []);
+		}
+	});
+	diffProperty('checked', properties, ({ checked: current }, { checked: next }) => {
+		if ((current || next) && current !== next) {
+			icache.set('controlledCheckedNodes', next || []);
+		}
+	});
+	diffProperty(
+		'initialChecked',
+		properties,
+		({ initialChecked: current }, { initialChecked: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('checked', next || []);
+			}
+		}
+	);
+	diffProperty(
+		'initialExpanded',
+		properties,
+		({ initialExpanded: current }, { initialExpanded: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('expanded', next || []);
+			}
+		}
+	);
+
+	const { getOrRead, createOptions, isLoading, meta } = resource;
+	const {
+		checkable = false,
+		selectable = false,
+		onValue,
+		onCheck,
+		onExpand,
+		disabledNodes,
+		resource: { template },
+		parentSelection = false
+	} = properties();
+	const classes = theme.classes(css);
+	const defaultRenderer = (n: TreeNodeOption) => n.value;
+	const [itemRenderer] = children();
+
+	const activeNode = icache.get('activeNode');
+	const selectedNode = icache.get('selectedNode');
+	const controlledExpandedNodes = icache.get('controlledExpandedNodes');
+	const controlledCheckedNodes = icache.get('controlledCheckedNodes');
+	const expandedNodes = icache.getOrSet('expanded', []);
+	const checkedNodes = icache.getOrSet('checked', []);
+	const shouldFocus = focus.shouldFocus();
+
+	function activateNode(id: string) {
+		icache.set('activeNode', id);
+	}
+
+	function selectNode(id: string) {
+		icache.set('selectedNode', id);
+		onValue && onValue(id);
+	}
+
+	function checkNode(id: string, checked: boolean) {
+		if (!controlledCheckedNodes) {
+			if (checked) {
+				icache.set('checked', (currentChecked) => [...currentChecked, id]);
+			} else {
+				icache.set('checked', (currentChecked) =>
+					currentChecked ? currentChecked.filter((n) => n !== id) : []
+				);
+			}
+		}
+		onCheck && onCheck(icache.get('checked') || []);
```

Comment:
I don't think this is going to work because you're only going to update the `checked` if it's not controlled so the listener in the controlled scenario will only ever get the previous checked value.
How about you just set the icache without the check and then call the `onCheck` with the result. There's no need to check if it's controlled at this point as you'r already doing it in the next render function and thus deciding if you should use the icache value or the `checked` value you've been passed in properties. The same is true for expand / collapse functionality. You should always be updating the `icache` and calling back with the latest.

Reply:
If you set the icache here there would be no difference between `initialChecked` and `controlledCheckedNodes`. The way I pictured the difference between `initialChecked` and `controlledCheckedNodes` was that the user would be in control of the when the boxes get clicked, if you set the `icache` here the `Tree` would update the checkbox without the user explicitly saying so